### DISCUSSION TITLE
Heterogeneous data random split for GNN

### DISCUSSION THREAD

**Discussion by 1AngelUS** [pyg-team/pytorch_geometric#5749] [Q&A]
Hi,
I am implementing a Hetero GNN by following the following link:
https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/hetero_conv_dblp.py

But in my data there are some nodes that don't have any label. Could you please help me as to how to random split the data to form train and test masks and train and test sets respectively. 

Currently I am splitting it this way:
n = merc_x.shape[0]
idx0=list(y0[:,0]) #label 0 samples
idx1=list(y1[:,0]) #label 1 samples

random.shuffle(idx0)
random.shuffle(idx1)

train_mask_idx1 = idx1[:4000]
train_mask_idx0 = idx0[:600000]

But the model is giving very poor performance and I think it is because of this splitting strategy.

Thanks very much!

**Comment by rusty1s** [comment]
The poor performance might be due to the imbalance of labels, it looks like you have a ratio of 1:15.

To understand your code better, how is `y0` and `y1` defined?

**Reply by 1AngelUS** [reply]
Thank you for you response. I got these labels (y0 and y1) directly. I had not defined them but the team which provided me this data told me they are using some other unsupervised model to define these. From their model they are getting a rank for each sample and they are assuming it in class 1 if rank < 5K and class 0 if rank > 50K. In between all the samples are assumed to have no label.

Yes, you are correct it might be because of imbalance of labels. That's why I am using imbalance_sampler with neighborloader. But, when I am calculating the precision and recall for the entire dataset it seems acceptable (specifically for class 1: precision= 0.30, recall=0.98). But when I am calculating the same for the **top K predictions** (where k=100,500,5000), **for class 1: precision= 0.004, recall=1.0 (out of 5000 positive predictions (sorted by probability of class 1 in decreasing order) by my model, only 2 of them are actually positive rest all of them belong to class 0).** That's why I thought it might be because of split as the behavior on the entire dataset should be similar to the behavior on the top K predictions.

Further, I trained a lightGBM model on the initial features directly (to have a baseline model), and to my surprise it is performing way better than GNN (for class 1: precision= 0.20, recall=0.88, this is for whole dataset and for top k predictions also). 

Request you to please help me regarding this!
Thank you once again!

**Reply by rusty1s** [reply]
Sorry for the late reply. I can't spot any issues in your splitting strategy, so I assume there might be something wrong within the GNN. What happens if you swap the GNN out for a simple MLP? That should at least give you similar performance as the lightGBM model.

**Reply by 1AngelUS** [reply]
Thanks once again for your kind reply! 

I ran my model after replacing SAGEConv layers by MLP and the performance was similar to the lightGBM model as you also suggested. 
What shall I do to improve GNN model performance? 

I was thinking may be to train the GNN in unsupervised setting. Could you help me with some source for the same.

**Reply by rusty1s** [reply]
In that case, I assume there is something wrong with your GNN or your given graph structure. How are your nodes connected to each other? Does the same hold true when using other GNN layers such as `GCNConv` or `GATConv`?

**Reply by 1AngelUS** [reply]
It is bipartite graph so only one type of edges and its a dense network. 

I will surely try other GNNs as you suggested and will let you know the results. I would also like to try an unsupervised GNN. Could you please help me with code sample for that?

**Reply by rusty1s** [reply]
If you want to perform message passing in a bipartite graph, I assume you need both directions. Otherwise, one set of nodes will never receive any updates.

**Reply by 1AngelUS** [reply]
Thank you for pointing this out. That could be an issue, I will change my code accordingly now.
